### PR TITLE
stats: Add API Scope::scopeFromStatName and use it to skip some symbol table locks during routing recompute from control-plane update.

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Adds an API for constructing scopes from StatName without holding a string representation, which helps avoid taking a lock. It was previously thought that scopes were not created on the hot-path, so this was presumed not to be necessary. But flame-graphs and code analysis suggest we can still contend on VirtualCluster creation, which involves making scopes.

As of now we are *not* trying to remove all creation of scopes from strings, but that is something we may want to consider as a follow-up.
Additional Description:
Risk Level: medium
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: include/envoy/stats/scope.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ class Scope {
    * @param name supplies the scope's namespace prefix.
    */
   virtual ScopePtr createScope(const std::string& name) PURE;
+  virtual ScopePtr scopeFromStatName(StatName name) PURE;
```
Comment (snowp) [inline include/envoy/stats/scope.h:48]: Doxygen docs?

File: source/common/stats/utility.h [hunk 2]
```diff
@@ -64,6 +64,24 @@ class Utility {
...
+   * tokens specified as DynamicName, a dynamic StatName will be created. See
+   * https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#dynamic-stat-tokens
+   * for more detail on why symbolic StatNames are preferred when possible.
+   *
+   * See also counterFromStatNames, which is slightly faster but does not allow
+   * passing DynamicName(string)s as names.
+   *
+   * @param scope The scope in which to create the counter.
+   * @param elements The vector of mixed DynamicName and StatName
+   * @param tags optionally specified tags.
```
Comment (snowp) [inline source/common/stats/utility.h:79]: remove?

File: source/common/stats/utility.h [hunk 3]
```diff
@@ -64,6 +64,24 @@ class Utility {
...
+   * @return A counter named using the joined elements.
+   */
+  static ScopePtr scopeFromElements(Scope& scope, const ElementVec& elements);
+  static ScopePtr scopeFromStatNames(Scope& scope, const StatNameVec& names);
```
Comment (snowp) [inline source/common/stats/utility.h:83]: Should this have its own doxygen comments?

**Review by snowp — DISMISSED** [review]
Looks right to me, just some comment comments :)

File: source/common/router/config_impl.h [hunk 4]
```diff
@@ -194,6 +194,12 @@ class VirtualHostImpl : public VirtualHost {
 private:
   enum class SslRequirements { None, ExternalOnly, All };
 
+  struct StatNameProvider {
+    StatNameProvider(absl::string_view name, Stats::SymbolTable& symbol_table)
+        : stat_name_storage_(std::make_unique<Stats::StatNameManagedStorage>(name, symbol_table)) {}
+    std::unique_ptr<Stats::StatNameManagedStorage> stat_name_storage_;
```
Comment (mattklein123) [inline source/common/router/config_impl.h:200]: Is there any reason this needs to be heap allocated within the struct? Can it be inline?

File: source/common/stats/thread_local_store.cc [hunk 5]
```diff
@@ -128,6 +128,14 @@ std::vector<CounterSharedPtr> ThreadLocalStoreImpl::counters() const {
 }
 
 ScopePtr ThreadLocalStoreImpl::createScope(const std::string& name) {
+  StatNameManagedStorage stat_name_storage(Utility::sanitizeStatsName(name), alloc_.symbolTable());
+  auto new_scope = std::make_unique<ScopeImpl>(*this, stat_name_storage.statName());
+  Thread::LockGuard lock(lock_);
+  scopes_.emplace(new_scope.get());
+  return new_scope;
```
Comment (mattklein123) [inline source/common/stats/thread_local_store.cc:135]: nit: can you just call `scopeFromStatName`?

File: source/common/stats/utility.h [hunk 6]
```diff
@@ -64,6 +64,24 @@ class Utility {
...
+   * https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#dynamic-stat-tokens
+   * for more detail on why symbolic StatNames are preferred when possible.
+   *
+   * See also counterFromStatNames, which is slightly faster but does not allow
+   * passing DynamicName(string)s as names.
+   *
+   * @param scope The scope in which to create the counter.
+   * @param elements The vector of mixed DynamicName and StatName
+   * @param tags optionally specified tags.
+   * @return A counter named using the joined elements.
```
Comment (mattklein123) [inline source/common/stats/utility.h:80]: I think these are out of date.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks generally LGTM with small comments. Should there be some explicit tests for the new functionality vs. the implicit testing we have here?

/wait

File: include/envoy/stats/scope.h (hunk 1 above)
Comment (jmarantz) [reply include/envoy/stats/scope.h:48]: done

File: source/common/router/config_impl.h (hunk 4 above)
Comment (jmarantz) [reply source/common/router/config_impl.h:200]: I thought there was some order-of-init issue but evidently not. Done.

File: source/common/stats/thread_local_store.cc (hunk 5 above)
Comment (jmarantz) [reply source/common/stats/thread_local_store.cc:135]: Yup, good catch, thanks.

File: source/common/stats/utility.h (hunk 2 above)
Comment (jmarantz) [reply source/common/stats/utility.h:79]: done

File: source/common/stats/utility.h (hunk 3 above)
Comment (jmarantz) [reply source/common/stats/utility.h:83]: done

**Review by mattklein123 — APPROVED** [review]
This LGTM though I would still prefer to have some explicit stats tests of the new behavior?

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**